### ISSUE TITLE
or not exists is ignored

### ISSUE THREAD

**Issue by andig** [influxdata/flux#3005]
Consider this query in Chronograf:

    from(bucket: "evcc")
    |> range(start: dashboardTime, stop: upperDashboardTime)
    |> filter(fn: (r) => r._measurement != "chargedEnergy")
    |> filter(fn: (r) => r.loadpoint == "Wallbe Eco" or not exists r.loadpoint)
  
It should return all measurements if specific tag or no tag. Instead, only measurements with tag are returned.

**Comment by nathanielc** [comment]
@andig Thanks for the report. We will take a look at this.

**Comment by markddavidoff** [comment]
+1 also seeing odd behaviour with "exists" in general, including this case in influxdb cloud 2.0.

**Comment by markddavidoff** [comment]
workaround:
```
    |> map(fn: (r) => ({r with keep_row: if exists r. loadpoint and r. loadpoint == true then false else true}))
    |> filter(fn: (r) => (r.keep_row ))
```

**Comment by andig** [comment]
Sorry, I don't get how the workaround is supposed to work. This doesn't:

    |> map(fn: (r) => ({r with keep_row: if exists r.loadpoint and r.loadpoint != ":loadpoint:" then false else true}))
    |> filter(fn: (r) => (r.keep_row))

@nathanielc any chance to get this solved instead of moved? It's insofar ugly as that you'll hardly notice something wrong unless you debug the actual query results.

**Comment by markddavidoff** [comment]
for me doing and if on the exits check and then swapping the returned values so it ends up returning not-exists worked as a workaround until not-exists is fixed

**Comment by andig** [comment]
With all due respect- there any chance of getting this fixed? This is an issue where query results are plain wrong and might go unnoticed.  Should Imho higher priority than feature requests.

**Comment by jsternberg** [comment]
@andig which version of the influxdb server are you using?

**Comment by andig** [comment]
> which version of the influxdb server are you using?

I'm running 1.8.2 using docker. Should I try something else? The workaround by @markddavidoff works but is highly unintuitive.

**Comment by jsternberg** [comment]
We're looking at this. It appears to be working properly in influxdb 2.0 and 1.8.2 hasn't been updated in awhile. We'll be figuring out a timeline to get 1.8.2 updated.

**Comment by jsternberg** [comment]
I suspect that this fix was never backported to 1.8: https://github.com/influxdata/influxdb/pull/15557

**Comment by jsternberg** [comment]
@andig can you double check and verify you are using 1.8.2? I ran it on my own machine doing the following and I am getting the appropriate result:

```
$ influx
> create database test1
> use test1
> insert m0,t0=a f0=1
> insert m0 f0=2
> exit
$ curl -XPOST -H 'Content-Type: application/vnd.flux' http://localhost:8086/api/v2/query -d 'from(bucket: "test1") |> range(start: -1h) |> filter(fn: (r) => r.t0 == "a" or not exists r.t0)'
#datatype,string,long,dateTime:RFC3339,dateTime:RFC3339,dateTime:RFC3339,double,string,string
#group,false,false,true,true,false,false,true,true
#default,_result,,,,,,,
,result,table,_start,_stop,_time,_value,_field,_measurement
,,0,2020-09-02T18:32:20.863634Z,2020-09-02T19:32:20.863634Z,2020-09-02T19:23:13.926561Z,2,f0,m0

#datatype,string,long,dateTime:RFC3339,dateTime:RFC3339,dateTime:RFC3339,double,string,string,string
#group,false,false,true,true,false,false,true,true,true
#default,_result,,,,,,,,
,result,table,_start,_stop,_time,_value,_field,_measurement,t0
,,1,2020-09-02T18:32:20.863634Z,2020-09-02T19:32:20.863634Z,2020-09-02T19:23:09.720449Z,1,f0,m0,a
```

It appears like `influx -type flux` might be broken so I used the direct curl query.

**Comment by jsternberg** [comment]
Also to clarify for this issue, the versions of flux between influx cloud and 1.x are very different. We haven't backported a flux release to 1.x in awhile so it is on v0.65.x while the cloud version is in the v0.82.x range. A major change to the language was merged directly after the last backport so these may be two separate problems.

**Comment by andig** [comment]
@jsternberg I can confirm your test. Took me a moment to figure, but it only works in the positive case- for whatever reason. Change `"a"` to `"b"` and you get an empty result:

    curl -XPOST -H 'Content-Type: application/vnd.flux' http://localhost:8086/api/v2/query -d \
    'from(bucket: "test1") |> range(start: -1h) |> filter(fn: (r) => r.t0 == "b" or not exists r.t0)'

**Comment by jsternberg** [comment]
I'm able to reproduce. Thanks.

**Comment by jsternberg** [comment]
This seems to be caused by a possible bug in the storage engine for 1.8. When it is attempting to learn the possible field names so it can create cursors, it doesn't see `r.t0 == ""` (which is what `not exists r.t0` gets translated to for storage queries) as valid series. When `r.t0 == "a"` is included, at least one series with the field key `_value` was matched so it ended up finding the data for the null value when it was creating the actual read cursor which does consider the above. When `r.t0 == "b"` is the only one used, then there is no series with that so it never finds any field keys that match the query because it isn't considering `r.t0 == ""` to be a valid check. This is because `r.t0 == ""` is true when a measurement has at least one series where it doesn't contain the tag key.

I'm looking to see how this can be fixed.

**Comment by jsternberg** [comment]
This can also be reproduced with just `filter(fn: (r) => not exists r.t0)`. The `or` isn't necessary.

**Comment by jsternberg** [comment]
It seems all negative comparisons are also ignored. The bug seems to affect _only_ when the storage service is using the fallback method of determining field keys. If the query thinks there is only one measurement, it uses a different code path that works properly.

The `OR` triggers the fallback method because this query does not pass the conditions put in here: https://github.com/influxdata/influxdb/blob/b222dfbe5ba8bc478959682874515e9eb721932c/services/storage/predicate_influxql.go#L28-L39

There is an `OR` so it uses the fallback method.

So another trigger is to write an `m0,t0=a` and `m0,t0=b` series and then doing `r.t0 != "a" or r.t0 == "c"`.

**Comment by andig** [comment]
Sounds like a lot of fun to fix :O

**Comment by jsternberg** [comment]
This bug affects v2 along with v1 so I created influxdata/influxdb#19566. After that is merged, I will backport it to 1.8.
